niceman/distributions/debian.py

```diff
@@ -368,93 +365,212 @@ def _find_all_sources(self):
-                        continue
-                    # Grab and name the source
-                    source = self._all_apt_sources[s]
-                    src_name = self._get_apt_source_name(source)
-                    source.name = src_name
-                    # Now add the source to our used sources
-                    self._apt_sources[src_name] = source
-                    # add the name for easy future lookup
-                    self._source_line_to_name_map[s] = src_name
-                # Look up and add the short name for the source
-                ver_dict[key].append(self._source_line_to_name_map[s])
-        return ver_dict
+    def _get_pkgs_arch_and_version(self, pkg_dicts):
+        # Convert package names to name:arch format
+        queries = []
+        for p in pkg_dicts:
+            # Use "dpkg -s pkg" to get the installed version and arch
+            architecture = p.get("architecture")
+            name = p.get("name")
+            queries.append(name if not architecture
+                           else "%s:%s" % (name, architecture))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
+        num_queries = max((_MAX_LEN_CMDLINE - 13) // (max_len + 1), 1)
+
+        cmd_results = []
+        # Execute dpkg -s, parse, and collate the output
+        for subqueries in (queries[pos:pos + num_queries]
+                           for pos in range(0, len(queries), num_queries)):
+            try:
+                out, _ = self._session.execute_command(
+                    ['dpkg', '-s'] + subqueries
+                )
+                out = utils.to_unicode(out, "utf-8")
+                # dpkg -s uses the same output as apt-cache show pkg
+                cmd_results += parse_apt_cache_show_pkgs_output(out)
+            except CommandError as _:
+                raise  # Currently raise any exception we encounter
+
+        # Turn dpkg -s results into a lookup table by package name
+        cmd_results = self.create_lookup_from_apt_cache_show(cmd_results)
+
+        for p in pkg_dicts:
+            # find the version table in the parsed command results for the
+            # current package
+            r = cmd_results.get("%s:%s" % (p.get("name"),
+                                             p.get("architecture")))
+            if not r:
+                r = cmd_results.get(p.get("name"))
+            if not r:
+                lgr.warning("Was unable to run dpkg -s for %s" %
+                            p.get("name"))
+                continue
+            p["architecture"] = r.get("Architecture")
+            p["version"] = r.get("Version")
+
+    @staticmethod
+    def create_lookup_from_apt_cache_show(cmd_results):
+        lookup_results = {}
+        for r in cmd_results:
+            lookup_results[r.get("Package")] = r
+            lookup_results["%s:%s" % (r.get("Package"),
+                                      r.get("architecture"))] = r
```

Comment:
getting lost in capitalizations  -- some times all upper case, some times all lower, and now mixed in ;)  any chance those all could be unified to e.g. lower case?

also note could be `"%(Package)s:%(Architecture)s" % r` or `"{Package}:{Architecture}".format(**r)` to avoid explicit `.get` and not end up with spurious None there (we better fail early in our testing than to discover those later)

Reply:
I updated the code to make our calls to get() explicit when we expect a value to be optional.  This allowed me to simplify some cases when I knew values should be set. I also used the shorter version to format the strings.

Reply:
I also made the parsing support function translate all keys to lower case to simplify the mapping.